jupyterhub/tests/test_roles.py

```diff
@@ -898,3 +941,52 @@ async def test_valid_names(name, valid):
     else:
         with pytest.raises(ValueError):
             roles._validate_role_name(name)
+
+
+@mark.role
+async def test_server_token_role(app):
+    user = add_user(app.db, app, name='test_user')
+    assert user.api_tokens == []
+    spawner = user.spawner
+    spawner.cmd = ['jupyterhub-singleuser']
+    await user.spawn()
+
+    server_token = spawner.api_token
+    orm_server_token = orm.APIToken.find(app.db, server_token)
+    assert orm_server_token
+
+    server_role = orm.Role.find(app.db, 'server')
+    token_role = orm.Role.find(app.db, 'token')
+    assert server_role in orm_server_token.roles
+    assert token_role not in orm_server_token.roles
+
+    assert orm_server_token.user.name == user.name
+    assert user.api_tokens == [orm_server_token]
+
+    await user.stop()
+
+
+@mark.role
+@mark.parametrize(
+    "token_role, response",
+    [
+        ('server', 200),
+        ('token', 200),
+        ('no_role', 403),
+    ],
+)
+async def test_server_posting_activity(app, token_role, response):
+    user = add_user(app.db, app, name='test_user')
+    if token_role == 'no_role':
+        api_token = user.new_api_token(roles=[])
+    else:
+        api_token = user.new_api_token(roles=[token_role])
+
+    r = await api_request(
```

Comment:
Can we also add a couple more tests to validate that the server scope doesn't grant more permissions than it should:

1. reading the owner's model from `GET /users` (should be able to read activity, but not other fields)
2. posting activity to another *existing* user should fail with 404, I think (maybe 403)

Reply:
Yes, good point, thanks.